File: ruby/Dice.rb
#encoding:utf-8

module Deepspace
    class Dice
        def initialize #son float constantes, menos generator
            @NHANGARSPROB = 0.25
            @NSHIELDSPROB = 0.25
            @NWEAPONSPROB = 0.33
            @FIRSTSHOTPROB = 0.5
            @generator = rand()
        end

        #Este método determina el número de hangares que recibirá una estación espacial al ser creada
        def initWithNHangars
            if @generator < @NHANGARSPROB   
                return 0
            else
                return 1
            end
        end

        #Este método determina el número de armas que recibirá una estación espacial al ser creada
        def initWithNWeapons
            aux = @generator
            if aux < @NWEAPONSPROB   
                return 1
            elsif aux < 2*@NWEAPONSPROB
                return 2
            else return 3
            end
        end

        #Este método determina el número de potenciadores de escudo que recibirá una estación espacial al ser creada
        def initWithNShields
            if @generator < @NSHIELDSPROB   
                return 0
            else
                return 1
            end
        end

        def whoStarts(nPlayers) #nPlayers es un int
            rand(nPlayers)
        end

        def firstShot
            if @generator < @FIRSTSHOTPROB   
                return GameCharacter::SPACESTATION
            else
                return GameCharacter::ENEMYSTARSHIP
            end
        end

        def spaceStationMoves(speed) #speed es un float, suponemos que entre 0 y 1. la probabilidad de moverse es mayor cuanto mas velocidad tenga
            if @generator < speed   
                return true #se movera
            else
                return false #no se movera
            end
        end

        def to_s
            return "NHangarsProb: #{@NHANGARSPROB}, NShieldsProb: #{@NSHIELDSPROB}, NWeaponsProb: #{@NWEAPONSPROB}, FirstShotProb: #{@FIRSTSHOTPROB}, Generator: #{@generator}"
        end
    end
end

File: java/Deepspace/src/deepspace/Dice.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.Random;

/**
 *
 * @author alvaro
 */
public class Dice {
    private final float NHANGARSPROB;
    private final float NSHIELDSPROB;
    private final float NWEAPONSPROB;
    private final float FIRSTSHOTPROB;
    private final float EXTRAEFFICIENCYPROB;
    private Random generator;
    
    Dice(){
        NHANGARSPROB = 0.25f;
        NSHIELDSPROB = 0.25f;
        NWEAPONSPROB = 0.33f;
        FIRSTSHOTPROB = 0.5f;
        EXTRAEFFICIENCYPROB = 0.8f;
        generator = new Random();
    }
    
    public int initWithNHangars(){
        if(generator.nextFloat() < NHANGARSPROB)
            return 0;
        else
            return 1;
    }
    
    public int initWithNWeapons(){
        float floatgenerado = generator.nextFloat();
        if(floatgenerado < NWEAPONSPROB)
            return 1;
        else if (floatgenerado < 2*NWEAPONSPROB)
                return 2;
            else return 3;
    }
    
    public boolean extraEfficiency(){
        return (generator.nextFloat() < EXTRAEFFICIENCYPROB);
    }
    
    public int initWithNShields(){
        if(generator.nextFloat() < NSHIELDSPROB)
            return 0;
        else
            return 1;
    }
    
    public int whoStarts(int nPlayers){
        return generator.nextInt(nPlayers);
    }
    
    public GameCharacter firstShot(){
        if(generator.nextFloat() < FIRSTSHOTPROB){
            return GameCharacter.SPACESTATION;
        }else return GameCharacter.ENEMYSTARSHIP;
    }
    
    public boolean spaceStationMoves(float speed){  //se asume que speed esta entre 0 y 1
        if(generator.nextFloat() < speed)
            return true;
        else return false;
    }
}
